.taskcluster.yml

```diff
@@ -5,158 +5,194 @@ version: 1
-        taskGroupId: ${decision_task_id}
-        created: {$fromNow: ''}
-        expires: ${expires_in}
-        deadline: {$fromNow: '2 hours'}
-        provisionerId: aws-provisioner-v1
-        workerType: github-worker
-        scopes:
-          - queue:create-task:aws-provisioner-v1/github-worker
-          - queue:scheduler-id:${scheduler_id}
-        payload:
-          maxRunTime: 7200
-          image: mozillamobile/android-components:1.15
-          env:
-            TASK_ID: ${decision_task_id}
-            SCHEDULER_ID: ${scheduler_id}
-            MOBILE_HEAD_REPOSITORY: ${repository}
-            MOBILE_HEAD_BRANCH: ${head_branch}
-            MOBILE_HEAD_REV: ${head_rev}
-            GITHUB_PULL_TITLE: ${pull_request_title}
-          command:
-            - /bin/bash
-            - --login
-            - -cx
-            - >-
-              cd ..
-              && git clone ${event.pull_request.head.repo.clone_url}
-              && cd reference-browser
-              && git config advice.detachedHead false
-              && git checkout ${event.pull_request.head.sha}
-              && python automation/taskcluster/decision_task.py
-          features:
-            taskclusterProxy: true
-          artifacts:
-            public/reports:
-              type: directory
-              path: /build/reference-browser/app/build/reports
-              expires: {$fromNow: '1 week'}
-        metadata:
-          name: Decision task
-          description: Building reference browser - triggered by a pull request.
-          owner: ${event.pull_request.user.login}@users.noreply.github.com
-          source: ${event.repository.url}
-  - $if: 'tasks_for == "cron"'
-    then:
-      $let:
-        decision_task_id: {$eval: as_slugid('decision_task')}
-        expires_in: {$fromNow: '1 year'}
-        repository: ${event.repository.html_url}
-        scheduler_id: focus-nightly-sched
-        is_mozilla_mobile_repo:
-          $eval: event.repository.html_url == 'https://github.com/mozilla-mobile/reference-browser'
-        command_staging_flag:
-          $if: event.repository.html_url == 'https://github.com/mozilla-mobile/reference-browser'
-          then: ''
-          else: '--staging'
-      in:
-        taskId: ${decision_task_id}
-        taskGroupId: ${decision_task_id}   # Must be explicit because of Chain of Trust
-        created: {$fromNow: ''}
-        deadline: {$fromNow: '2 hours'}
-        expires: ${expires_in}
-        schedulerId: ${scheduler_id}   # Must be explicit because of Chain of Trust
-        provisionerId: aws-provisioner-v1
-        # TODO Replace by mobile-3-decision (https://bugzilla.mozilla.org/show_bug.cgi?id=1512631)
-        workerType: gecko-focus   # This workerType has ChainOfTrust enabled
-        requires: all-completed   # Must be explicit because of Chain of Trust
-        priority: medium
-        retries: 5
-        scopes:
-          $flatten:
+        default_task_definition:
```

Comment:
In my opinion, the bits in `default_task_definition` are already general items: no matter the type, a decision task must have a`taskId`, a `taskGroupId`, and also a `workerType`, a `priority`, a `repository` to clone from, etc. 

I agree, the first pattern is not ideal when the example is a few lines long. It becomes more relevant when the `if` statement spans over a hundred lines, I think. For example, if you want to understand what are the priorities defined, you need to:

1. find the first occurrence of `priority`
2. scroll up a little bit to find what it was for (say for PRs)
3. remember the value
4. Ctrl+F to the next occurrence
5. Repeat step 1 to 4 until you covered all cases.

=> it takes a lot of steps, and you need to mentally store each value between each step. On the other hand, you get to know the full logic, by looking at these 9 lines: https://github.com/mozilla-mobile/reference-browser/pull/579/files#diff-ac0229d1171c0983b27003af4c7441a5R52. Then, when you look at the PR section (to answer your 3rd point), you just get what is uncommon compared to the other types of events. Otherwise, if we want to compare how different a push event and a PR event, we need to:

1. For each parameter of a PR, Ctrl+F until you find the equivalent in the push section
2. For each parameter of the push section, Ctrl+F until you find the equivalent in the PR section
3. Mentally remember the similarities *and* the differences + mentally remember what was missing in the other. 

Regarding your second point, android-components handles 4 types of events: https://github.com/mozilla-mobile/android-components/blob/5e32f657de33ae635c113b1756fd356df72518a6/.taskcluster.yml. You can easily see that PRs and pushes [are very similar](https://github.com/mozilla-mobile/android-components/blob/master/.taskcluster.yml#L116-L151), and so do releases and cron. 

So far, one big drawback I can think of is: if I want to debug something within the PR definition, then a couple dict definitions to check, and to see which one overrides the other. Though, I believe this case is rarer and modifying values that affect several types of graphs is more common and more error-prone. Therefore, I think we should optimize for this case, and that what this patterns intends.

Reply:
> In my opinion, the bits in `default_task_definition` are already general items: no matter the type, a decision task must have a`taskId`, a `taskGroupId`, and also a `workerType`, a `priority`, a `repository` to clone from, etc.

I feel like we're conflating separate goals: 
1. Grouping custom properties in the task type section, while having shared properties (even _with different values_) defined above
2. Only sharing properties and defining them above if they have the same value. Declare everything task-type-specific in the task type section

My preference is to have everything "type-specific" specific in the task type section, and I believe that if a property's value changes for a task type, then it's a type-specific property.

> I agree, the first pattern is not ideal when the example is a few lines long. It becomes more relevant when the `if` statement spans over a hundred lines, I think. For example, if you want to understand what are the priorities defined, you need to: ...

I agree that it's easy to see all permutations of a property if you define them all together, but I'm not sure that this is worth the drawbacks of this approach. One such example is the converse: what I want to look up all defined properties for a task? I can't even CTRL-F for the `tasks_for`, because it might be defined in an `else`.

(Related: the `else` doesn't make sense at-a-glance. What task type is that for? You can resolve this by forcing the use of `$if` for all the different states of the declaration, but now you're enforcing boilerplate and are increasing the burden on code review)

> Then, when you look at the PR section (to answer your 3rd point), you just get what is uncommon compared to the other types of events.

That's the thing: you only see _some_ of the changed properties when you look at the PR section (like `payload`). You _can't_ tell that properties like `priority` or `scheduler_id` have changed.

> Otherwise, if we want to compare how different a push event and a PR event, we need to: ...

I'd argue that it's tougher to compare a push event vs a PR event when we declare our properties generally up top. Not only do we need to directly compare the task type sections for both (regardless of this discussion), we _also_ have to scroll through _all_ the property declarations and watch where they differ between push and PR. Meanwhile, if we defined everything type-specific in the task type section, then we _just_ have to compare the two type sections.

> You can easily see that PRs and pushes [are very similar](https://github.com/mozilla-mobile/android-components/blob/master/.taskcluster.yml#L116-L151), and so do releases and cron.

I'm 100% on-board with declaring shared properties when the values are the same. Even in cases where a subset of task types share the properties, just do the declaration in the same context (close by), just like how `pr_or_push_parameters` was done in the [android-components example you linked](https://github.com/mozilla-mobile/android-components/blob/master/.taskcluster.yml#L116-L126). They're shared between PR and push events, and they're declared within the `tasks_for in ["github-pull-request", "github-push"]` block. Awesome!

> Though, I believe [working on specific task type definitions] is rarer and modifying values that affect several types of graphs is more common and more error-prone. Therefore, I think we should optimize for this case, and that what this patterns intends.

Yes, if we need to change the worker type, we'll only need to change it in one place. However, we've got a bit of an inheritance-y system going on: we declare a top-level generic task, then extend it for each task type. This comes with the downsides of this style:
* When a change is needed to a shared value that only affects a subset of the "extend-ers", your base logic becomes infested with `$if`/`else` logic based on who is extending the parent, which adds noise and complexity
* It's a lot easier to change a value and _accidentally_ affect a task graph that you didn't mean to. This increases burden on code review, and when issues get introduced, the cause isn't immediately obvious.

Needing to apply a value change to multiple task types might take an extra minute, but it's explicit and has way less surprises, e.g.: "PR and push should now use the worker type 'super-worker', so I'll change the PR task type and push task type". Besides, we read code much more frequently than we change it. If it takes use an extra minute to change, but understanding how different tasks are created is easier, then I'd opt for the readable solution.

Reply:
Johan and I talked via Vidyo earlier today: though "extension" and "inheritance" might have non-negligible maintenance costs associated with it, it allows greater code-reuse than we'd be able to accomplish with composition due to the limits JSON-e enforces (it's not a full programming language, so we can't create the same composition abstractions).

TL;DR: conversation is resolved, let's merge as-is